common/network/queues/concurrent_queue.c.inc

```diff
@@ -0,0 +1,110 @@
+    size = self->size;                                     \
+    lock_handle_unlock(&self->lock);                       \
+    return size == 0;                                      \
+  }
+
+#define DEFINE_FUNCTION_SIZE_CONCURRENT_QUEUE_OF(TYPE)     \
+  size_t CONCURRENT_QUEUE_OF_##TYPE##_SIZE(                \
+      struct concurrent_queue_of_##TYPE##_s *const self) { \
+    size_t size;                                           \
+    if (self == NULL) return 0;                            \
+    lock_handle_lock(&self->lock);                         \
+    size = self->size;                                     \
+    lock_handle_unlock(&self->lock);                       \
+    return size;                                           \
+  }
+
+#define DEFINE_FUNCTION_PUSH_CONCURRENT_QUEUE_OF(TYPE)                      \
+  concurrent_queue_status_t CONCURRENT_QUEUE_OF_##TYPE##_PUSH(              \
+      struct concurrent_queue_of_##TYPE##_s *const self, TYPE const data) { \
+    concurrent_queue_node_of_##TYPE##_t *node;                              \
+    if (self == NULL) return CONCURRENT_QUEUE_NULL_SELF;                    \
+    node = calloc(1, sizeof(*node));                                        \
+    if (node == NULL) return CONCURRENT_QUEUE_ALLOC_FAILED;                 \
+    node->data = data;                                                      \
+    lock_handle_lock(&self->lock);                                          \
+    if (self->front == NULL) {                                              \
+      self->front = node;                                                   \
+      self->back = node;                                                    \
+    } else {                                                                \
+      self->back->next = node;                                              \
+      self->back = node;                                                    \
+    }                                                                       \
+    self->size++;                                                           \
+    lock_handle_unlock(&self->lock);                                        \
+    cond_handle_signal(&self->cond);                                        \
+    return CONCURRENT_QUEUE_SUCCESS;                                        \
+  }
+
+#define DEFINE_FUNCTION_POP_CONCURRENT_QUEUE_OF(TYPE)                        \
+  concurrent_queue_status_t CONCURRENT_QUEUE_OF_##TYPE##_POP(                \
+      struct concurrent_queue_of_##TYPE##_s *const self, TYPE *const data) { \
+    if (self == NULL) return CONCURRENT_QUEUE_NULL_SELF;                     \
+    if (data == NULL) return CONCURRENT_QUEUE_NULL_DATA;                     \
+    lock_handle_lock(&self->lock);                                           \
+    while (self->size == 0) {                                                \
+      cond_handle_wait(&self->cond, &self->lock);                            \
```

Comment:
this won't work though?
you hold the lock, so `push` won't be able to acquire the lock.

Reply:
The action to wait on a condition variable actually releases the associated lock

> The pthread_cond_wait() function atomically blocks the current thread waiting on the condition variable specified by cond, and releases the mutex
>      specified by mutex.  The waiting thread unblocks only after another thread calls pthread_cond_signal(3), or pthread_cond_broadcast(3) with the same
>      condition variable, and the current thread reacquires the lock on mutex.